maglev: Delete map if previous M's do not match

brb opened the pull request:
It was reported by a user, that when the M param
(`--bpf-lb-maglev-table-size`) changes, then the verifier is failing to
load the program on the 5.4 kernel with the following error:

    invalid access to map value, value_size=32762 off=65520 size=2 R0
    max value is outside of the array range processed 507 insns (limit
    1000000) max_states_per_insn 1 total_states 34 peak_states 34
    mark_read 17

Fix this by recreating the map when we detect that the M has changed.

Fix: #14251
Reported-by: Han Zhou <hzhou8@ebay.com>

brb committed:
maglev: Delete map if previous M do not match

It was reported by a user, that when the M param
(--bpf-lb-maglev-table-size) changes, then the verifier is failing to
load the program on the 5.4 kernel with the following error:

    invalid access to map value, value_size=32762 off=65520 size=2 R0
    max value is outside of the array range processed 507 insns (limit
    1000000) max_states_per_insn 1 total_states 34 peak_states 34
    mark_read 17

Fix this by recreating the map when we detect that the M has changed.

Reported-by: Han Zhou <hzhou8@ebay.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on pkg/maps/lbmap/maglev.go:
```diff
@@ -60,6 +68,55 @@ func InitMaglevMaps(ipv4 bool, ipv6 bool) error {
+
+	if deleteMap {
+		log.WithField(logfields.BPFMapName, mapName).
+			Info("Deleting Maglev outer map due to possibly different M")
```
Why "possibly"? Doesn't the above check guarantee that M changed?

borkmann replied:
Maybe `err == io.EOF` is meant here but agree with Paul

brb replied:
The map is deleted in one of the following cases:

1. There was entry in the outer map, and from that entry we can derive the previous M which is != the current M.
2. The outer map was empty, so we could not derive the previous M. Therefore, we don't know whether it was different.

Because of the 2. case I added "possibly" to the log msg.

pchaigno replied:
It's really a nit, but: `due to different M or empty map`?

brb replied:
Let's wait until the CI tests have finished, then I will push the suggestion.

pchaigno reviewed (approved):
I would have expected that check to happen in `cilium-map-migrate` since that's where we currently check for property mismatches. But given we want to favor an all-Golang codebase in the long term, it probably makes sense to implement this logic in Go :+1: 

LGTM!

borkmann reviewed (approved):
lgtm

brb committed:
version: Move GetKernelVersion

Move the GetKernelVersion() from pkg/datapath/linux to pkg/version so
that it can be included by pkg/maps/lbmap which is included by
pkg/datapath/linux. Otherwise, we have a circular dependency.

Also, make the function to return an error instead of calling log.Fatal,
as it's caller's responsibility to decide what to do with an error.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add skip for maglev map tests

Unfortunately, we run the privileged unit tests on the 4.9 kernel which
does not have the map-in-map support. Therefore, add a skip based on the
kernel version so that the CI does not fail.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
